Codespaces: Use the host name from the logged in server for commands

jkeech opened the pull request:
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->
Fixes https://github.com/github/codespaces/issues/14711

* Refactor the Codespaces root command configuration to follow the same pattern as other commands
    * Expose a `NewCmdCodespace` function to build the command and invoke it from `pkg/cmd/root/root.go`
    * Remove the lazy HTTP client initialization wrapper that existed in `pkg/cmd/root/root.go`. Push the HTTP client downstream into command runtime in `internal/codespaces/api/api.go` (which keeps it lazy, but is simpler and follows the pattern of the other commands)
* Use the logged in auth context to provide the API base url. Previously it would respect the `GITHUB_API_URL` environment variable if set, otherwise fallback to a hardcoded `https://api.github.com`. Now it uses the following precedence:
    * `GITHUB_API_URL` environment variable
    * Logged in auth context
    * Fallback to hardcoded `https://api.github.com`
 * Similarly for the server url, use the following precedence:
    * `GITHUB_SERVER_URL` environment variable
    * Logged in auth context
    * Fallback to hardcoded `https://github.com`
 * Remove dead code in the codespaces api package relating to the `vscsAPI` property

jkeech committed:
Use the host name from the logged in server for codespace commands

jkeech committed:
Fix existing tests

jkeech committed:
Add tests for server url configuration

jkeech commented on internal/codespaces/api/api.go:
```diff
@@ -42,13 +43,13 @@ import (
 )
 
 const (
-	githubServer = "https://github.com"
```
`githubServer` and `vscsAPI` values were not in use anymore so I cleaned those up. I believe they were dead code from earlier versions of the Codespaces API.

jkeech replied:
I ended up needing to add the server url back in to support the `--web` flows of list/create, but the `vscsAPI` is no longer needed.

cliAutomation commented:
Hi! Thanks for the pull request. Please ensure that this change is linked to an issue by mentioning an issue number in the description of the pull request. If this pull request would close the issue, please put the word 'Fixes' before the issue number somewhere in the pull request body. If this is a tiny change like fixing a typo, feel free to ignore this message.

samcoe commented on internal/codespaces/api/api_test.go:
```diff
@@ -126,13 +131,99 @@ func createFakeCreateEndpointServer(t *testing.T, wantStatus int) *httptest.Serv
+}
+
+func TestNewApiServerUrl_dotcomConfig(t *testing.T) {
+	os.Setenv("GITHUB_API_URL", "")
```
```suggestion
	t.Setenv("GITHUB_API_URL", "")
```

Let's make sure to use `t.Setenv` in tests rather than `os.Setenv`.

jkeech replied:
Thanks for pointing this out! I wasn't aware of this method, but I agree it's much better for tests after looking it up. I actually modeled this after a few others tests that were calling `os.Setenv`. Would you like me to update all of them in this PR?

https://github.com/search?q=repo%3Acli%2Fcli+os.setEnv+path%3A**%2F*_test.go&type=code

jkeech replied:
I went ahead and updated them all and kept the non-codespaces ones in another commit so that it's easy to rebase it out if you want to keep that separate.

samcoe commented on internal/codespaces/api/api.go:
```diff
@@ -42,13 +43,13 @@ import (
-	githubServer = "https://github.com"
-	githubAPI    = "https://api.github.com"
-	vscsAPI      = "https://online.visualstudio.com"
+	defaultApiUrl = "https://api.github.com"
```
```suggestion
	defaultAPIURL = "https://api.github.com"
```
Small naming nit per Go convention.

samcoe reviewed (commented):
@jkeech Code looks good, thanks for adding tests! I left a couple small inline comments. Let me know if any of them are unclear.

There is one larger issues with this change that I noticed. There are numerous places in the `cs` commands that have the `https://github.com/codespaces` URL hardcoded that will need to be changed to also properly handle a configurable hostname.

jkeech committed:
Rename defaultApiUrl to defaultAPIURL and comment cleanup

jkeech committed:
Switch to t.setEnv in codespaces api tests

jkeech committed:
Switch to t.Setenv in other tests

jkeech committed:
Support custom server in web flows for list and create

jkeech commented:
@samcoe, I've addressed all of your feedback in the latest iteration, including fixing the `--web` scenarios on `gh cs list` and `gh cs create`. Thanks for pointing those out!

dmgardiner25 commented on internal/codespaces/api/api.go:
```diff
@@ -60,31 +60,40 @@ const (
-		serverURL = githubServer
-	}
+func New(f *cmdutil.Factory) *API {
+	apiURL := os.Getenv("GITHUB_API_URL")
```
Could we make this and the server URL consts like the URLs above?

jkeech replied:
Looking at other usages of `os.Getenv` in this repo, almost all of them use an inlined string rather than a separate const. I'd rather follow the conventions of the rest of the codebase. I also personally find it easier to read the code with fewer layers of indirection.

https://github.com/search?q=repo%3Acli%2Fcli%20os.Getenv&type=code

dmgardiner25 reviewed (approved):
Looks good, just one small comment!

samcoe commented on internal/codespaces/api/api.go:
```diff
@@ -97,6 +106,11 @@ type User struct {
 }
 
+// GetServerURL returns the server url (not the API url), such as https://github.com
+func (a *API) GetServerURL() string {
```
Small naming nit, but what do you think of just calling this `ServerURL()`? Having `Get` in the name of an API client method makes me think it is going to make a GET HTTP request. I don't feel strongly about this, just a thought.

jkeech replied:
Sure thing. I don't have much of an opinion on it either, and I'm not very familiar with Go naming conventions so I'm happy to update it based on your recommendation. Updated in the latest commit.

samcoe reviewed (approved):
Code LGTM, thanks for making the requested changes! I left one small non-blocking idea in the code.

jkeech committed:
Rename GetServerURL() to ServerURL()
